### PR TITLE
Feature/add generator for submodels

### PR DESCRIPTION
**Pull request by andreas-schilling**

### CODE REVIEW

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 1]
```diff
@@ -192,7 +170,7 @@ public Environment visitBase( final ModelElement base, final Context context ) {
    public Environment visitAspect( final Aspect aspect, Context context ) {
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:170]: `context` should be final, method arguments should not be reassigned

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 2]
```diff
@@ -496,29 +391,31 @@ private EmbeddedDataSpecification extractEmbeddedDataSpecification( final Aspect
    private DataSpecificationIec61360 extractDataSpecificationContent( final Property property ) {
       final List<LangStringDefinitionTypeIec61360> definitions = property.getCharacteristic().stream().flatMap( characteristic ->
                   characteristic.getDescriptions().stream() )
-            .map( this::mapDefinitionIec61360 )
+            .map( LangStringMapper.DEFINITION::map )
             .collect( Collectors.toList() );
 
       final List<LangStringPreferredNameTypeIec61360> preferredNames = property.getPreferredNames().size() > 0 ?
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:397]: Replace `.size() > 0` with `.isEmpty()`

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 3]
```diff
@@ -558,102 +450,156 @@ public Environment visitCharacteristic(
...
+            .flatMap( dataType -> handleEitherField( "left", either.getLeft(), property, context ) )
+            .ifPresent( submodelElements::add );
+      either.getRight().getDataType()
+            .flatMap( dataType -> handleEitherField( "right", either.getRight(), property, context ) )
+            .ifPresent( submodelElements::add );
+
+      final SubmodelElementList eitherSubModelElements =
             new DefaultSubmodelElementList.Builder()
                   .idShort( ID_PREFIX + either.getName() )
                   .typeValueListElement( AASSubmodelElements.DATA_ELEMENT ) // TODO check if more specific type info is rqujried
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:551]: Can this TODO be addressed? Also, there is a typo (a different one than in the TODO comment above :-))

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 4]
```diff
@@ -558,102 +450,156 @@ public Environment visitCharacteristic(
...
+      Optional<SubmodelElement> result = Optional.empty();
+      if ( context.getModelingKind().equals( ModellingKind.INSTANCE ) ) {
+         final var fieldProperty = createProperty( eitherProperty.getMetaModelVersion(), field, fieldCharacteristic );
+         context.setProperty( fieldProperty );
+         if ( context.getRawPropertyValue().isPresent() ) {
+            result = Optional.ofNullable( decideOnMapping( fieldCharacteristic.getDataType().get(), context.getProperty(), context ) );
+         }
+         context.getPropertyResult();
+      } else {
+         result = Optional.ofNullable( decideOnMapping( fieldCharacteristic.getDataType().get(), context.getProperty(), context ) );
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:582]: Optional .get() should not be called without .map()/.isPresent()

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+
+/**
+ * Base interface for any class that can map a property to a {@link SubmodelElement}.
+ *
+ * @param <T> the concrete type of {@link SubmodelElement} the implementing mapper produces
+ */
+public interface PropertyMapper<T extends SubmodelElement> {
+   static final String UNKNOWN_TYPE = "Unknown";
+
+   static final String UNKNOWN_EXAMPLE = UNKNOWN_TYPE;
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java:43]: Remove redundant `static final`

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGenerator.java [hunk 6]
```diff
@@ -71,20 +87,56 @@ public void generateAasJsonFile(
...
+         serializer.write( out, mergedEnvironment );
+         return out;
+      } catch ( final SerializationException e ) {
+         throw new IOException( e );
+      }
+   }
+
+   private Environment mergeEnvironments( final Map<Aspect, Environment> aspectEnvironments ) {
+      final Submodel submodel = new DefaultSubmodel.Builder().build();
+      final Environment environment = new DefaultEnvironment.Builder()
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGenerator.java:117]: Replace assignment with `return`

File: core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 7]
```diff
@@ -278,7 +360,18 @@ private Environment getAssetAdministrationShellFromAspect( final TestAspect test
...
+   }
+
+   private Environment getAssetAdministrationShellFromAspectWithData( final TestAspect testAspect )
+         throws DeserializationException, IOException, SAXException {
+      final Aspect aspect = loadAspect( testAspect );
+      final JsonNode aspectData = loadPayload( testAspect );
+      final ByteArrayOutputStream out = generator.generateXmlOutput( Map.of( aspect, aspectData ) );
+      final var data = out.toByteArray();
+      final var result = loadAASX( data, testAspect );
+      //new AasXmlValidator().validateXml( data );
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java:373]: Can/should this be removed?

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/Context.java [hunk 8]
```diff
@@ -12,26 +12,68 @@
...
+import com.fasterxml.jackson.databind.JsonNode;
 import org.eclipse.esmf.metamodel.Property;
 
+
+/**
+ * Contains and tracks the context while the {@link AspectModelAASVisitor} traverses the metamodel.
+ */
 public class Context {
 
    Environment environment;
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/Context.java:40]: Make all those fields that can be `final`

File: core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java [hunk 9]
```diff
@@ -12,14 +12,19 @@
...
 package org.eclipse.esmf.aspectmodel.aas;
 
+import static org.assertj.core.api.Assertions.assertThat;
 import static org.junit.jupiter.api.Assertions.assertEquals;
 import static org.junit.jupiter.api.Assertions.assertTrue;
 
 import java.io.ByteArrayInputStream;
 import java.io.ByteArrayOutputStream;
 import java.io.IOException;
+import java.io.InputStream;
```
Comment (atextor) [inline core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java:22]: Remove unused imports

**Review by atextor — CHANGES_REQUESTED** [review]

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java (hunk 2 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:397]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 10]
```diff
@@ -558,102 +450,156 @@ public Environment visitCharacteristic(
...
+                  final var values = StreamSupport.stream( arrayNode.spliterator(), false )
+                        .map( n -> {
+                           context.iterate( property );
+                           return decideOnMapping( property, context );
+                        } )
+                        .toList();
+                  context.finishIteration( property );
+                  return values;
+               }
+            } ).orElseGet( () -> List.of() );
```
[the remark this answers is not in the document]
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:529]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java (hunk 3 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:551]: Same, DATA_ELEMENT should be fine.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+
+import com.google.common.collect.ImmutableMap;
+
+/**
+ * Base interface for any class that can map a property to a {@link SubmodelElement}.
+ *
+ * @param <T> the concrete type of {@link SubmodelElement} the implementing mapper produces
+ */
+public interface PropertyMapper<T extends SubmodelElement> {
+   static final String UNKNOWN_TYPE = "Unknown";
```
[the remark this answers is not in the document]
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java:41]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java (hunk 5 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/PropertyMapper.java:43]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGenerator.java (hunk 6 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGenerator.java:117]: Done.

File: core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java (hunk 7 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java:373]: Removed.

File: core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java (hunk 9 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/test/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASGeneratorTest.java:22]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/Context.java (hunk 8 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/Context.java:40]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java [hunk 12]
```diff
@@ -558,102 +450,156 @@ public Environment visitCharacteristic(
...
+    * @return the submodel element for the {@code Either} field
+    */
+   private Optional<SubmodelElement> handleEitherField( final String field, final Characteristic fieldCharacteristic,
+         final Property eitherProperty, final Context context ) {
+      Optional<SubmodelElement> result = Optional.empty();
+      if ( context.getModelingKind().equals( ModellingKind.INSTANCE ) ) {
+         final var fieldProperty = createProperty( eitherProperty.getMetaModelVersion(), field, fieldCharacteristic );
+         context.setProperty( fieldProperty );
+         if ( context.getRawPropertyValue().isPresent() ) {
+            result = Optional.ofNullable( decideOnMapping( fieldCharacteristic.getDataType().get(), context.getProperty(), context ) );
```
[the remark this answers is not in the document]
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:578]: Done.

File: core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java (hunk 1 above)
Comment (andreas-schilling) [reply core/esmf-aspect-model-aas-generator/src/main/java/org/eclipse/esmf/aspectmodel/aas/AspectModelAASVisitor.java:170]: Done.

**Pull request merged**